### PR TITLE
 Enable asynchronous deprovision in TSB

### PR DESCRIPTION
**Pull request by jim-minter**

### CODE REVIEW

File: pkg/templateservicebroker/servicebroker/deprovision.go [hunk 1]
```diff
@@ -56,31 +55,13 @@ func (b *Broker) Deprovision(u user.Info, instanceID string) *api.Response {
...
-		return api.Forbidden(err)
-	}
-
-	err = b.kc.Core().Secrets(brokerTemplateInstance.Spec.Secret.Namespace).Delete(brokerTemplateInstance.Spec.Secret.Name, metav1.NewPreconditionDeleteOptions(string(brokerTemplateInstance.Spec.Secret.UID)))
-	if err != nil && !kerrors.IsNotFound(err) {
-		return api.InternalServerError(err)
-	}
-
 	// The OSB API requires this function to be idempotent (restartable).  If
 	// any actual change was made, per the spec, StatusOK is returned, else
```
Comment (bparees) [inline pkg/templateservicebroker/servicebroker/deprovision.go]: comment needs to be updated.

File: test/extended/templates/templateservicebroker_e2e.go [hunk 2]
```diff
@@ -259,25 +280,78 @@ var _ = g.Describe("[Conformance][templates] templateservicebroker end-to-end te
 
 	deprovision := func() {
 		g.By("deprovisioning a service")
-		err := brokercli.Deprovision(context.Background(), cliUser, instanceID)
+		err := cli.TemplateClient().Template().Templates(cli.Namespace()).Delete(privatetemplate.Name, &metav1.DeleteOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
```
Comment (bparees) [inline test/extended/templates/templateservicebroker_e2e.go]: what's the point of deleting the template here?

File: test/extended/templates/templateservicebroker_e2e.go [hunk 3]
```diff
@@ -259,25 +280,78 @@ var _ = g.Describe("[Conformance][templates] templateservicebroker end-to-end te
...
+
+		err = brokercli.Deprovision(context.Background(), cliUser, instanceID)
```
Comment (bparees) [inline test/extended/templates/templateservicebroker_e2e.go]: doesn't this need to wait until lastoperation returns success?  (should have a test for that anyway)

File: test/extended/templates/templateservicebroker_e2e.go (hunk 2 above)
Comment (jim-minter) [reply test/extended/templates/templateservicebroker_e2e.go]: What I've implemented is that the test explicitly checks that the namespace is left empty (inasmuch as this is possible).  We manually put that template in the namespace earlier (in the BeforeEach), therefore we must manually remove it for the test to succeed.

File: test/extended/templates/templateservicebroker_e2e.go (hunk 3 above)
Comment (jim-minter) [reply test/extended/templates/templateservicebroker_e2e.go]: brokercli (already) handles that internally.  What additional test are you looking for?

File: pkg/templateservicebroker/servicebroker/deprovision.go (hunk 1 above)
Comment (jim-minter) [reply pkg/templateservicebroker/servicebroker/deprovision.go]: fixed, thanks.

File: test/extended/templates/templateservicebroker_e2e.go (hunk 3 above)
Comment (bparees) [reply test/extended/templates/templateservicebroker_e2e.go]: ok, I didn't dig into brokercli explicitly.  I see now that it already handles both sync and async responses (I was assuming it only handled sync and would need changes to add support for async)

**Comment by bparees** [comment]
i think this is lgtm but will revisit after https://github.com/openshift/origin/pull/16808 merges, i don't want to merge this beforehand since it's got a cherrypicked commit.

**Comment by jim-minter** [comment]
@bparees #16808 has now merged.  I rebased this and repushed - just 1 commit showing now.

**Pull request merged**